diff --git a/node_modules/@react-native/codegen/lib/generators/RNCodegen.js b/node_modules/@react-native/codegen/lib/generators/RNCodegen.js
index e5f9a14..5918e14 100644
--- a/node_modules/@react-native/codegen/lib/generators/RNCodegen.js
+++ b/node_modules/@react-native/codegen/lib/generators/RNCodegen.js
@@ -16,6 +16,9 @@ TODO:
 - ViewConfigs should spread in View's valid attributes
 */
 const schemaValidator = require('../SchemaValidator.js');
+const {
+  generateSupportedApplePlatformsMacro,
+} = require('./components/ComponentsProviderUtils');
 const generateComponentDescriptorCpp = require('./components/GenerateComponentDescriptorCpp.js');
 const generateComponentDescriptorH = require('./components/GenerateComponentDescriptorH.js');
 const generateComponentHObjCpp = require('./components/GenerateComponentHObjCpp.js');
@@ -263,4 +266,5 @@ module.exports = {
     }
     return result.value;
   },
+  generateSupportedApplePlatformsMacro,
 };
