diff --git a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
index dc0df16..401912c 100644
--- a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
+++ b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
@@ -4,11 +4,13 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  *
+ * @flow strict-local
  * @format
  */
 
 'use strict';
 
+const {getCodegen} = require('../codegen-utils');
 const {TEMPLATES_FOLDER_PATH} = require('./constants');
 const {
   codegenLog,
@@ -28,7 +30,11 @@ const THIRD_PARTY_COMPONENTS_MM_TEMPLATE_PATH = path.join(
   'RCTThirdPartyComponentsProviderMM.template',
 );
 
-function generateRCTThirdPartyComponents(libraries, outputDir) {
+function generateRCTThirdPartyComponents(
+  libraries /*: $ReadOnlyArray<$FlowFixMe> */,
+  librariesSupportedApplePlatforms /*: $FlowFixMe */,
+  outputDir /*: string */,
+) {
   fs.mkdirSync(outputDir, {recursive: true});
   // Generate Header File
   codegenLog('Generating RCTThirdPartyComponentsProvider.h');
@@ -41,7 +47,8 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
   codegenLog(`Generated artifact: ${finalPathH}`);
 
   codegenLog('Generating RCTThirdPartyComponentsProvider.mm');
-  let componentsInLibraries = {};
+  let componentsInLibraries = {} /*:: as {[string]: Array<$FlowFixMe>} */;
+  let componentsSupportedApplePlatforms = {};
 
   const componentLibraries = libraries.filter(({config, libraryPath}) => {
     if (isReactNativeCoreLibrary(config.name) || config.type === 'modules') {
@@ -50,7 +57,7 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
     return true;
   });
 
-  const librariesToCrawl = {};
+  const librariesToCrawl = {} /*:: as {[string]: $FlowFixMe} */;
 
   // Using new API explicitly or not using any config field to define components.
   const componentLibrariesUsingNewApi = [];
@@ -71,9 +78,12 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
   componentLibrariesUsingOldApi.forEach(library => {
     const {config, libraryPath} = library;
     const libraryName = JSON.parse(
-      fs.readFileSync(path.join(libraryPath, 'package.json')),
+      fs.readFileSync(path.join(libraryPath, 'package.json'), 'utf8'),
     ).name;
 
+    componentsSupportedApplePlatforms[libraryName] =
+      librariesSupportedApplePlatforms[config.name];
+
     librariesToCrawl[libraryName] = library;
 
     const componentsProvider = config.ios?.componentProvider;
@@ -130,14 +140,11 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
     componentsInLibraries[libraryName] = componentsMapping;
   });
 
-  const thirdPartyComponentsMapping = Object.keys(componentsInLibraries)
-    .flatMap(library => {
-      const components = componentsInLibraries[library];
-      return components.map(({componentName, className}) => {
-        return `\t\t@"${componentName}": NSClassFromString(@"${className}"), // ${library}`;
-      });
-    })
-    .join('\n');
+  const thirdPartyComponentsMapping = _generateComponentRegistry(
+    componentsInLibraries,
+    componentsSupportedApplePlatforms,
+  );
+
   // Generate implementation file
   const templateMM = fs
     .readFileSync(THIRD_PARTY_COMPONENTS_MM_TEMPLATE_PATH, 'utf8')
@@ -152,7 +159,10 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
 
 // Given a path, return the paths of all the files with extension .mm in
 // the path dir and all its subdirectories.
-function findFilesWithExtension(filePath, extension) {
+function findFilesWithExtension(
+  filePath /*: string */,
+  extension /*: string */,
+) /*: Array<string> */ {
   const files = [];
   const dir = fs.readdirSync(filePath);
   dir.forEach(file => {
@@ -184,7 +194,7 @@ function findFilesWithExtension(filePath, extension) {
 
 // Given a filepath, read the file and look for a string that starts with 'Class<RCTComponentViewProtocol> '
 // and ends with 'Cls(void)'. Return the string between the two.
-function findRCTComponentViewProtocolClass(filepath) {
+function findRCTComponentViewProtocolClass(filepath /*: string */) {
   const fileContent = fs.readFileSync(filepath, 'utf8');
   const regex = /Class<RCTComponentViewProtocol> (.*)Cls\(/;
   const match = fileContent.match(regex);
@@ -218,6 +228,51 @@ function findRCTComponentViewProtocolClass(filepath) {
   return null;
 }
 
+function _generateComponentRegistry(
+  libraryComponentsMap /*: $FlowFixMe */,
+  platformSupportByLibrary /*: $FlowFixMe */,
+) /*: string */ {
+  if (!libraryComponentsMap || Object.keys(libraryComponentsMap).length === 0) {
+    return '';
+  }
+
+  const generatedComponentMappings = Object.keys(libraryComponentsMap).flatMap(
+    library => {
+      const components = libraryComponentsMap[library];
+
+      const generateSupportedApplePlatformsMacro =
+        getCodegen().generateSupportedApplePlatformsMacro;
+
+      const componentMappings = components
+        .map(({componentName, className}) => {
+          if (!className || !componentName) {
+            return null;
+          }
+
+          return `\t\t@"${componentName}": NSClassFromString(@"${className}"), // ${library}`;
+        })
+        .filter(item => item !== null);
+
+      if (componentMappings.length === 0) {
+        return ([] /*: Array<string> */);
+      }
+
+      return generateSupportedApplePlatformsMacro(
+        componentMappings.join('\n'),
+        platformSupportByLibrary?.[library],
+      );
+    },
+  );
+
+  const sanitizedComponentMappings = generatedComponentMappings
+    .map(line => line.replace(/[\r\n]+$/, ''))
+    .filter(line => line.length > 0);
+
+  return sanitizedComponentMappings.join('\n');
+}
+
 module.exports = {
   generateRCTThirdPartyComponents,
+  // exported for testing purposes only:
+  _generateComponentRegistry,
 };
diff --git a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
index b062ad6..9b34283 100644
--- a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
+++ b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
@@ -4,6 +4,7 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  *
+ * @flow strict-local
  * @format
  */
 
@@ -58,11 +59,11 @@ const path = require('path');
  * @throws If it can't find a cli for the CodeGen.
  */
 function execute(
-  projectRoot,
-  targetPlatform,
-  baseOutputPath,
-  source,
-  runReactNativeCodegen = true,
+  projectRoot /*: string */,
+  targetPlatform /*: string */,
+  baseOutputPath /*: string */,
+  source /*: string */,
+  runReactNativeCodegen /*: boolean */ = true,
 ) {
   try {
     codegenLog(`Analyzing ${path.join(projectRoot, 'package.json')}`);
@@ -120,6 +121,7 @@ function execute(
         platform,
       );
 
+      let librariesSupportedApplePlatforms = {};
       if (runReactNativeCodegen) {
         const schemaInfos = generateSchemaInfos(libraries);
         generateNativeCode(
@@ -130,11 +132,23 @@ function execute(
           pkgJsonIncludesGeneratedCode(pkgJson),
           platform,
         );
+
+        librariesSupportedApplePlatforms = schemaInfos.reduce(
+          (acc, {library, supportedApplePlatforms}) => {
+            acc[library.config.name] = supportedApplePlatforms;
+            return acc;
+          },
+          {},
+        );
       }
 
       if (source === 'app' && platform !== 'android') {
         // These components are only required by apps, not by libraries and are Apple specific.
-        generateRCTThirdPartyComponents(libraries, outputPath);
+        generateRCTThirdPartyComponents(
+          libraries,
+          librariesSupportedApplePlatforms,
+          outputPath,
+        );
         generateRCTModuleProviders(projectRoot, pkgJson, libraries, outputPath);
         generateCustomURLHandlers(libraries, outputPath);
         generateUnstableModulesRequiringMainQueueSetupProvider(
@@ -161,7 +175,10 @@ function execute(
   return;
 }
 
-function readOutputDirFromPkgJson(pkgJson, platform) {
+function readOutputDirFromPkgJson(
+  pkgJson /*: $FlowFixMe */,
+  platform /*: string */,
+) {
   const codegenConfig = pkgJson.codegenConfig;
   if (codegenConfig == null || typeof codegenConfig !== 'object') {
     return null;
@@ -179,12 +196,19 @@ function readOutputDirFromPkgJson(pkgJson, platform) {
   return null;
 }
 
-function computeOutputPath(projectRoot, baseOutputPath, pkgJson, platform) {
+function computeOutputPath(
+  projectRoot /*: string */,
+  baseOutputPath /*: string */,
+  pkgJson /*: $FlowFixMe */,
+  platform /*: string */,
+) {
   if (baseOutputPath == null) {
     const outputDirFromPkgJson = readOutputDirFromPkgJson(pkgJson, platform);
     if (outputDirFromPkgJson != null) {
+      // $FlowFixMe[reassign-const]
       baseOutputPath = path.join(projectRoot, outputDirFromPkgJson);
     } else {
+      // $FlowFixMe[reassign-const]
       baseOutputPath = projectRoot;
     }
   }
@@ -201,7 +225,7 @@ function computeOutputPath(projectRoot, baseOutputPath, pkgJson, platform) {
   return baseOutputPath;
 }
 
-function defaultOutputPathForAndroid(baseOutputPath) {
+function defaultOutputPathForAndroid(baseOutputPath /*: string */) {
   return path.join(
     baseOutputPath,
     'android',
@@ -213,11 +237,14 @@ function defaultOutputPathForAndroid(baseOutputPath) {
   );
 }
 
-function defaultOutputPathForIOS(baseOutputPath) {
+function defaultOutputPathForIOS(baseOutputPath /*: string */) {
   return path.join(baseOutputPath, 'build', 'generated', 'ios');
 }
 
-function mustGenerateNativeCode(includeLibraryPath, schemaInfo) {
+function mustGenerateNativeCode(
+  includeLibraryPath /*: string */,
+  schemaInfo /*: $FlowFixMe */,
+) {
   // If library's 'codegenConfig' sets 'includesGeneratedCode' to 'true',
   // then we assume that native code is shipped with the library,
   // and we don't need to generate it.
