diff --git a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
index f7e74b6..401912c 100644
--- a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
+++ b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
@@ -10,6 +10,7 @@
 
 'use strict';
 
+const {getCodegen} = require('../codegen-utils');
 const {TEMPLATES_FOLDER_PATH} = require('./constants');
 const {
   codegenLog,
@@ -31,6 +32,7 @@ const THIRD_PARTY_COMPONENTS_MM_TEMPLATE_PATH = path.join(
 
 function generateRCTThirdPartyComponents(
   libraries /*: $ReadOnlyArray<$FlowFixMe> */,
+  librariesSupportedApplePlatforms /*: $FlowFixMe */,
   outputDir /*: string */,
 ) {
   fs.mkdirSync(outputDir, {recursive: true});
@@ -46,6 +48,7 @@ function generateRCTThirdPartyComponents(
 
   codegenLog('Generating RCTThirdPartyComponentsProvider.mm');
   let componentsInLibraries = {} /*:: as {[string]: Array<$FlowFixMe>} */;
+  let componentsSupportedApplePlatforms = {};
 
   const componentLibraries = libraries.filter(({config, libraryPath}) => {
     if (isReactNativeCoreLibrary(config.name) || config.type === 'modules') {
@@ -56,19 +59,34 @@ function generateRCTThirdPartyComponents(
 
   const librariesToCrawl = {} /*:: as {[string]: $FlowFixMe} */;
 
+  // Using new API explicitly or not using any config field to define components.
+  const componentLibrariesUsingNewApi = [];
+  const componentLibrariesUsingOldApi = [];
+
+  for (const library of componentLibraries) {
+    if (
+      library.config.ios?.components ||
+      !library.config.ios?.componentProvider
+    ) {
+      componentLibrariesUsingNewApi.push(library);
+    } else {
+      componentLibrariesUsingOldApi.push(library);
+    }
+  }
+
   // Old API
-  componentLibraries.forEach(library => {
+  componentLibrariesUsingOldApi.forEach(library => {
     const {config, libraryPath} = library;
     const libraryName = JSON.parse(
       fs.readFileSync(path.join(libraryPath, 'package.json'), 'utf8'),
     ).name;
 
+    componentsSupportedApplePlatforms[libraryName] =
+      librariesSupportedApplePlatforms[config.name];
+
     librariesToCrawl[libraryName] = library;
 
     const componentsProvider = config.ios?.componentProvider;
-    if (!componentsProvider) {
-      return;
-    }
 
     delete librariesToCrawl[libraryName];
     componentsInLibraries[libraryName] =
@@ -83,7 +101,7 @@ function generateRCTThirdPartyComponents(
   });
 
   // New API
-  const iosAnnotations = parseiOSAnnotations(componentLibraries);
+  const iosAnnotations = parseiOSAnnotations(componentLibrariesUsingNewApi);
   for (const [libraryName, annotationMap] of Object.entries(iosAnnotations)) {
     const {library, components} = annotationMap;
     librariesToCrawl[libraryName] = library;
@@ -122,14 +140,11 @@ function generateRCTThirdPartyComponents(
     componentsInLibraries[libraryName] = componentsMapping;
   });
 
-  const thirdPartyComponentsMapping = Object.keys(componentsInLibraries)
-    .flatMap(library => {
-      const components = componentsInLibraries[library];
-      return components.map(({componentName, className}) => {
-        return `\t\t@"${componentName}": NSClassFromString(@"${className}"), // ${library}`;
-      });
-    })
-    .join('\n');
+  const thirdPartyComponentsMapping = _generateComponentRegistry(
+    componentsInLibraries,
+    componentsSupportedApplePlatforms,
+  );
+
   // Generate implementation file
   const templateMM = fs
     .readFileSync(THIRD_PARTY_COMPONENTS_MM_TEMPLATE_PATH, 'utf8')
@@ -213,6 +228,51 @@ function findRCTComponentViewProtocolClass(filepath /*: string */) {
   return null;
 }
 
+function _generateComponentRegistry(
+  libraryComponentsMap /*: $FlowFixMe */,
+  platformSupportByLibrary /*: $FlowFixMe */,
+) /*: string */ {
+  if (!libraryComponentsMap || Object.keys(libraryComponentsMap).length === 0) {
+    return '';
+  }
+
+  const generatedComponentMappings = Object.keys(libraryComponentsMap).flatMap(
+    library => {
+      const components = libraryComponentsMap[library];
+
+      const generateSupportedApplePlatformsMacro =
+        getCodegen().generateSupportedApplePlatformsMacro;
+
+      const componentMappings = components
+        .map(({componentName, className}) => {
+          if (!className || !componentName) {
+            return null;
+          }
+
+          return `\t\t@"${componentName}": NSClassFromString(@"${className}"), // ${library}`;
+        })
+        .filter(item => item !== null);
+
+      if (componentMappings.length === 0) {
+        return ([] /*: Array<string> */);
+      }
+
+      return generateSupportedApplePlatformsMacro(
+        componentMappings.join('\n'),
+        platformSupportByLibrary?.[library],
+      );
+    },
+  );
+
+  const sanitizedComponentMappings = generatedComponentMappings
+    .map(line => line.replace(/[\r\n]+$/, ''))
+    .filter(line => line.length > 0);
+
+  return sanitizedComponentMappings.join('\n');
+}
+
 module.exports = {
   generateRCTThirdPartyComponents,
+  // exported for testing purposes only:
+  _generateComponentRegistry,
 };
diff --git a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
index 0022734..9b34283 100644
--- a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
+++ b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
@@ -121,6 +121,7 @@ function execute(
         platform,
       );
 
+      let librariesSupportedApplePlatforms = {};
       if (runReactNativeCodegen) {
         const schemaInfos = generateSchemaInfos(libraries);
         generateNativeCode(
@@ -131,11 +132,23 @@ function execute(
           pkgJsonIncludesGeneratedCode(pkgJson),
           platform,
         );
+
+        librariesSupportedApplePlatforms = schemaInfos.reduce(
+          (acc, {library, supportedApplePlatforms}) => {
+            acc[library.config.name] = supportedApplePlatforms;
+            return acc;
+          },
+          {},
+        );
       }
 
       if (source === 'app' && platform !== 'android') {
         // These components are only required by apps, not by libraries and are Apple specific.
-        generateRCTThirdPartyComponents(libraries, outputPath);
+        generateRCTThirdPartyComponents(
+          libraries,
+          librariesSupportedApplePlatforms,
+          outputPath,
+        );
         generateRCTModuleProviders(projectRoot, pkgJson, libraries, outputPath);
         generateCustomURLHandlers(libraries, outputPath);
         generateUnstableModulesRequiringMainQueueSetupProvider(
