diff --git a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
index 53669fa..9e0ae20 100644
--- a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
+++ b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/generateRCTThirdPartyComponents.js
@@ -15,6 +15,7 @@ const {
   isReactNativeCoreLibrary,
   parseiOSAnnotations,
 } = require('./utils');
+const {getCodeGen} = require('../codegen-utils');
 const fs = require('fs');
 const path = require('path');
 
@@ -28,7 +29,11 @@ const THIRD_PARTY_COMPONENTS_MM_TEMPLATE_PATH = path.join(
   'RCTThirdPartyComponentsProviderMM.template',
 );
 
-function generateRCTThirdPartyComponents(libraries, outputDir) {
+function generateRCTThirdPartyComponents(
+  libraries,
+  librariesSupportedApplePlatforms,
+  outputDir,
+) {
   fs.mkdirSync(outputDir, {recursive: true});
   // Generate Header File
   codegenLog('Generating RCTThirdPartyComponentsProvider.h');
@@ -42,6 +47,7 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
 
   codegenLog('Generating RCTThirdPartyComponentsProvider.mm');
   let componentsInLibraries = {};
+  let componentsSupportedApplePlatforms = {};
 
   const componentLibraries = libraries.filter(({config, libraryPath}) => {
     if (isReactNativeCoreLibrary(config.name) || config.type === 'modules') {
@@ -59,6 +65,9 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
       fs.readFileSync(path.join(libraryPath, 'package.json')),
     ).name;
 
+    componentsSupportedApplePlatforms[libraryName] =
+      librariesSupportedApplePlatforms[config.name];
+
     librariesToCrawl[libraryName] = library;
 
     const componentsProvider = config.ios?.componentProvider;
@@ -118,14 +127,8 @@ function generateRCTThirdPartyComponents(libraries, outputDir) {
     componentsInLibraries[libraryName] = componentsMapping;
   });
 
-  const thirdPartyComponentsMapping = Object.keys(componentsInLibraries)
-    .flatMap(library => {
-      const components = componentsInLibraries[library];
-      return components.map(({componentName, className}) => {
-        return `\t\t@"${componentName}": NSClassFromString(@"${className}"), // ${library}`;
-      });
-    })
-    .join('\n');
+  const thirdPartyComponentsMapping = _generateComponentRegistry(componentsInLibraries, componentsSupportedApplePlatforms);
+
   // Generate implementation file
   const templateMM = fs
     .readFileSync(THIRD_PARTY_COMPONENTS_MM_TEMPLATE_PATH, 'utf8')
@@ -206,6 +209,46 @@ function findRCTComponentViewProtocolClass(filepath) {
   return null;
 }
 
+function _generateComponentRegistry(libraryComponentsMap, platformSupportByLibrary) {
+  if (!libraryComponentsMap || Object.keys(libraryComponentsMap).length === 0) {
+    return '';
+  }
+
+  const generateSupportedApplePlatformsMacro = getCodeGen().generateSupportedApplePlatformsMacro;
+
+  const generatedComponentMappings = Object.keys(libraryComponentsMap)
+    .flatMap(library => {
+      const components = libraryComponentsMap[library];
+
+      const componentMappings = components.map(
+        ({ componentName, className }) => {
+          if (!className || !componentName) {
+            return null;
+          }
+
+          return `\t\t@"${componentName}": NSClassFromString(@"${className}"), // ${library}`;
+        },
+      ).filter(item => item !== null);
+
+      if (componentMappings.length === 0) {
+        return [];
+      }
+
+      return generateSupportedApplePlatformsMacro(
+        componentMappings.join('\n'),
+        platformSupportByLibrary?.[library]
+      );
+    });
+
+  const sanitizedComponentMappings = generatedComponentMappings
+    .map(line => line.replace(/[\r\n]+$/, ''))
+    .filter(line => line.length > 0);
+
+  return sanitizedComponentMappings.join('\n');
+}
+
 module.exports = {
   generateRCTThirdPartyComponents,
+  // exported for testing purposes only:
+  _generateComponentRegistry,
 };
diff --git a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
index 4b9ca0e..0d1c139 100644
--- a/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
+++ b/node_modules/react-native/scripts/codegen/generate-artifacts-executor/index.js
@@ -101,6 +101,7 @@ function execute(
         platform,
       );
 
+      let librariesSupportedApplePlatforms = {};
       if (runReactNativeCodegen) {
         const schemaInfos = generateSchemaInfos(libraries);
         generateNativeCode(
@@ -111,11 +112,23 @@ function execute(
           pkgJsonIncludesGeneratedCode(pkgJson),
           platform,
         );
+
+        librariesSupportedApplePlatforms = schemaInfos.reduce(
+          (acc, {library, supportedApplePlatforms}) => {
+            acc[library.config.name] = supportedApplePlatforms;
+            return acc;
+          },
+          {},
+        );
       }
 
       if (source === 'app') {
         // These components are only required by apps, not by libraries
-        generateRCTThirdPartyComponents(libraries, outputPath);
+        generateRCTThirdPartyComponents(
+          libraries,
+          librariesSupportedApplePlatforms,
+          outputPath,
+        );
         generateRCTModuleProviders(projectRoot, pkgJson, libraries, outputPath);
         generateCustomURLHandlers(libraries, outputPath);
         generateUnstableModulesRequiringMainQueueSetupProvider(
